src/processor/operator/table_scan/ftable_scan_function.cpp

```diff
@@ -17,23 +18,23 @@ struct FTableScanMorsel {
         : startTupleIdx{startTupleIdx}, numTuples{numTuples} {}
 };
 
-struct FTableScanSharedState final : public TableFuncSharedState {
-    std::mutex mtx;
+struct FTableScanSharedState final : public function::BaseScanSharedState {
     std::shared_ptr<FactorizedTable> table;
     uint64_t morselSize;
     common::offset_t nextTupleIdx;
 
     FTableScanSharedState(std::shared_ptr<FactorizedTable> table, uint64_t morselSize)
-        : TableFuncSharedState{}, table{std::move(table)}, morselSize{morselSize}, nextTupleIdx{0} {
-    }
+        : BaseScanSharedState{}, table{std::move(table)}, morselSize{morselSize}, nextTupleIdx{0} {}
 
     FTableScanMorsel getMorsel() {
-        std::unique_lock lck{mtx};
+        std::unique_lock lck{lock};
         auto numTuplesToScan = std::min(morselSize, table->getNumTuples() - nextTupleIdx);
         auto morsel = FTableScanMorsel(nextTupleIdx, numTuplesToScan);
         nextTupleIdx += numTuplesToScan;
         return morsel;
     }
+
+    uint64_t getNumRows() const override { return table->getTotalNumFlatTuples(); }
```

Comment:
I think this will slows down the performance of caller.

I'm pretty sure 99% when we access numRows we are NOT accessing the number of flat tuples (only iterator need this I believe). So either return numRows here or don't expose this interface.